src/Core/OrganizationFeatures/AuthorizationHandlers/GroupAuthHandler.cs

```diff
@@ -0,0 +1,33 @@
+﻿using Bit.Core.Context;
+using Bit.Core.Entities;
+using Bit.Core.Enums;
+using Microsoft.AspNetCore.Authorization;
+
+namespace Bit.Core.OrganizationFeatures.AuthorizationHandlers;
+
+class GroupAuthHandler : AuthorizationHandler<GroupOperationRequirement, Group>
+{
+    private readonly ICurrentContext _currentContext;
+
+    public GroupAuthHandler(ICurrentContext currentContext)
+    {
+        _currentContext = currentContext;
+    }
+
+    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
```

Comment:
I seriously dislike that this is a single method interface. I know it's so that we can have cool Attribute decorators, but this is nasty.

Please establish the pattern than this method is just a switch out to other methods that handle each `GroupOperationRequirement` individually.

Reply:
`switch` only works for constant expressions but I've used `if` for this (not sure how literal you were being). It was the structure I had in mind but you're right we should set it up early.

Reply:
dang. Forgot about C#'s less than stellar enum and switch. That just means you have to have a bit more boilerplate, but it's still doable without confusing boolean logic or if/else if/else chains:

<details>
<summary>I wrote some code for you</summary>

<details>
<summary>option 1</summary>

Basically, this makes `GroupOperations` a hashable entity by implementing `IEquatable<T>` and uses that to form a `Dictionary` of how to handle each Operation. What you get out the other end is basically a separated switch statement. 

We can, of course improve this with some reflection, but for now I think this is enough to get started and be understandable.

```csharp
// GroupAuthorizationHandler.cs

using Bit.Core.Context;
using Bit.Core.Entities;
using Bit.Core.Enums;
using Microsoft.AspNetCore.Authorization;

namespace Bit.Core.OrganizationFeatures.AuthorizationHandlers;

class GroupAuthorizationHandler : AuthorizationHandler<GroupOperationRequirement, Group>
{
    private readonly ICurrentContext _currentContext;

    private readonly Dictionary<GroupOperationRequirement, Action<AuthorizationHandlerContext, GroupOperationRequirement, Group>> _handlers;

    public GroupAuthorizationHandler(ICurrentContext currentContext)
    {
        _currentContext = currentContext;

        _handlers = new() {
            { GroupOperations.Create, canCreate },
            { GroupOperations.Read, canRead },
            { GroupOperations.Update, canUpdate },
            { GroupOperations.Delete, canDelete },
            { GroupOperations.AddUser, canAddUser },
            { GroupOperations.DeleteUser, canDeleteUser },
        };
    }

    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
        GroupOperationRequirement requirement,
        Group resource)
    {
        if (_handlers.TryGetValue(requirement, out var handler))
        {
            handler(context, requirement, resource);
        }

        await Task.CompletedTask;
    }

    private void canCreate(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void canRead(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void canUpdate(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void canDelete(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void canAddUser(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void canDeleteUser(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void CanManageGroups(AuthorizationHandlerContext context,
        GroupOperationRequirement requirement,
        Group resource)
    {
        // TODO: providers need to be included in the claims
        var org = _currentContext.GetOrganization(resource.OrganizationId);
        var canAccess = org.Type == OrganizationUserType.Owner ||
                        org.Type == OrganizationUserType.Admin ||
                        org.Permissions.ManageGroups;

        if (canAccess)
        {
            context.Succeed(requirement);
        }
    }
}
```

```csharp
// GroupOperations.cs

using Microsoft.AspNetCore.Authorization.Infrastructure;

namespace Bit.Core.OrganizationFeatures.AuthorizationHandlers;

public class GroupOperationRequirement : OperationAuthorizationRequirement, IEquatable<GroupOperationRequirement>
{
    public bool Equals(GroupOperationRequirement other)
    {
        return other != null && Name == other.Name;
    }

    public override bool Equals(object obj) => Equals(obj as GroupOperationRequirement);

    public override int GetHashCode() => Name.GetHashCode();
}

public static class GroupOperations
{
    // Operations on the Group object itself
    public static readonly GroupOperationRequirement Create = new() { Name = nameof(Create) };
    public static readonly GroupOperationRequirement Read = new() { Name = nameof(Read) };
    public static readonly GroupOperationRequirement Update = new() { Name = nameof(Update) };
    public static readonly GroupOperationRequirement Delete = new() { Name = nameof(Delete) };

    // Operations on Group-User associations 
    public static readonly GroupOperationRequirement AddUser = new() { Name = nameof(AddUser) };
    public static readonly GroupOperationRequirement DeleteUser = new() { Name = nameof(DeleteUser) };
}
```
</details>
<details>
<summary>option 2</summary>

This one works by encoding the correct handler method on the Operation itself. I kind of like this one better, but it's not really type safe, uses reflection, and kind of forms a tautological dev loop -- which comes first, the operation or the method???? We can probably create a Util method to generate the `Handler` Expression-bodied member.

```csharp
// GroupAuthorizationHandler.cs

using Bit.Core.Context;
using Bit.Core.Entities;
using Bit.Core.Enums;
using Microsoft.AspNetCore.Authorization;

namespace Bit.Core.OrganizationFeatures.AuthorizationHandlers;

public class GroupAuthorizationHandler : AuthorizationHandler<GroupOperationRequirement, Group>
{
    private readonly ICurrentContext _currentContext;

    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
        GroupOperationRequirement requirement,
        Group resource)
    {
        requirement.Handler(this, context, requirement, resource);

        await Task.CompletedTask;
    }

    public void canCreate(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    public void canRead(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    public void canUpdate(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    public void canDelete(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    public void canAddUser(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    public void canDeleteUser(AuthorizationHandlerContext context, GroupOperationRequirement requirement, Group resource)
    {
        CanManageGroups(context, requirement, resource);
    }

    private void CanManageGroups(AuthorizationHandlerContext context,
        GroupOperationRequirement requirement,
        Group resource)
    {
        // TODO: providers need to be included in the claims
        var org = _currentContext.GetOrganization(resource.OrganizationId);
        var canAccess = org.Type == OrganizationUserType.Owner ||
                        org.Type == OrganizationUserType.Admin ||
                        org.Permissions.ManageGroups;

        if (canAccess)
        {
            context.Succeed(requirement);
        }
    }
}
```

```csharp
// GroupOperations.cs

using Bit.Core.Entities;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Authorization.Infrastructure;

namespace Bit.Core.OrganizationFeatures.AuthorizationHandlers;

public class GroupOperationRequirement : OperationAuthorizationRequirement
{
    public Action<GroupAuthorizationHandler, AuthorizationHandlerContext, GroupOperationRequirement, Group> Handler =>
        (Action<GroupAuthorizationHandler, AuthorizationHandlerContext, GroupOperationRequirement, Group>)Delegate.CreateDelegate(typeof(Action<GroupAuthorizationHandler, AuthorizationHandlerContext, GroupOperationRequirement, Group>), typeof(GroupAuthorizationHandler).GetMethod(AuthorizationHandlerMethodName));
    public string AuthorizationHandlerMethodName { get; init; }
}

public static class GroupOperations
{
    // Operations on the Group object itself
    public static readonly GroupOperationRequirement Create = new() { Name = nameof(Create), AuthorizationHandlerMethodName = nameof(GroupAuthorizationHandler.canCreate) };
    public static readonly GroupOperationRequirement Read = new() { Name = nameof(Read), AuthorizationHandlerMethodName = nameof(GroupAuthorizationHandler.canRead) };
    public static readonly GroupOperationRequirement Update = new() { Name = nameof(Update), AuthorizationHandlerMethodName = nameof(GroupAuthorizationHandler.canUpdate) };
    public static readonly GroupOperationRequirement Delete = new() { Name = nameof(Delete), AuthorizationHandlerMethodName = nameof(GroupAuthorizationHandler.canDelete) };

    // Operations on Group-User associations 
    public static readonly GroupOperationRequirement AddUser = new() { Name = nameof(AddUser), AuthorizationHandlerMethodName = nameof(GroupAuthorizationHandler.canAddUser) };
    public static readonly GroupOperationRequirement DeleteUser = new() { Name = nameof(DeleteUser), AuthorizationHandlerMethodName = nameof(GroupAuthorizationHandler.canDeleteUser) };
}
``` 

I took the time to test this out on an analog for the cost we pay for reflection. If we were to reflect on each invocation, it's pretty bad, but the way I wrote it, we actually only reflect the one time on instantiation of the static Operation ([Check out this blog ](https://codeblog.jonskeet.uk/2008/08/09/making-reflection-fly-and-exploring-delegates/)). It's not totally free like in that blog post (might be an artifact of my simple invoked method), but the cost is quite small.

|            Method |        Mean |     Error |    StdDev |      Median |
|------------------ |------------:|----------:|----------:|------------:|
|        DirectCall |   0.0002 ns | 0.0008 ns | 0.0008 ns |   0.0000 ns |
|    ReflectionCall | 121.9399 ns | 0.7676 ns | 0.6804 ns | 122.0550 ns |
| SingleReflectCall *this example* |   5.0782 ns | 0.0525 ns | 0.0438 ns |   5.0878 ns |

</details>
</details>

Curious on everyone's thoughts. Usually people don't like it when I do this kind of functional or reflection stuff.

I would expect a meta-test that ensures that the functional map `_handlers` is instantiated with each type of `GroupOperation` or that the reflection used in the `GroupOperation` works in addition to typical functionality tests for `GroupOperation` equalities and permission handling.

Reply:
Resolved using @Thomas-Avery's approach in his PR, which allows us to use switch statements after all :)